### DISCUSSION TITLE
High-level difference between SimpleTuner and the training script in huggingface

### DISCUSSION THREAD

**Discussion by zjysteven** [bghira/SimpleTuner#251] [Q&A] [answered]
Hi,

Thank you for such great efforts. I saw that there is a sample finetuning script for SDXL on huggingface (https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_sdxl.py), and you were involved in the discussion there too. I was just wondering that on a high level what's the difference between that script and SimpleTuner. Can SimpleTuner be considered an extension for that script with much more detailed considerations?

Thanks

**Comment by bghira** [comment, marked as answer]
this script has many features than the huggingface trainer (that my team created) submitted to the Diffusers project.

there's probably too many differences to trivially list in one place, but the main ones:

* better tracking of dataset training progress, eg. no over-sampling of data unintentionally
* aspect bucketing for multi-resolution support in the final model
* better caching for VAE / text embeds, stored on disk and not recalculated on each start
* better multi-GPU support, but still no real multi-node support
* DeepSpeed integration that is tested and working properly
* better crop conditional usage for SDXL models, not just using `(0, 0)` (the new/proposed SDXL Turbo trainer fixes that for its needs but didn't fix the issue for teh XL trainer)
* more statistics being tracked during training, eg. train vs validation luminance
* ability to train directly from an S3 bucket, including the VAE cache entries

**Reply by zjysteven** [reply]
Thank you @bghira for the detailed comments. A very impressive project too!

**Reply by bghira** [reply]
thank you, i always appreciate and am surprised by the generous words people leave on their reports and questions here.

**Comment by bghira** [comment]
in short, simpletuner is optimised for **high quality training results** while the huggingface script is a **bare minimum to get results**.

this trainer works with millions of images without issue, whereas the diffusers script will choke and struggle with more than a few thousand images.

**Reply by BuildBackBuehler** [reply]
You should definitely put that in as the tag phrase at the top of your Read Me! If I wasn't getting so obsessive over training thanks to my Mac setup, I doubt I'd stumble upon your project (or care enough to read through the overview). But if I saw that its superior results/better handling of important parameters like multi-resolution images I'd pick it over a barebones script any day! 

I started with Kohya's scripts, found them clunky, hard-to-follow and unoptimized. Then Diffuser's script. I don't know if I'm doing something wrong, but it took me 2 damn hours to figure out how to convert a single-file safetensor model, I prefer unzip though because that at least differentiates it from all the straightforward format-to-format conversions. 

Y'anyways, thanks for yours and Bhenzesi (sp) work. I wish I'd seen this, I was using Bhenzesi's Diffusers to get some MPS/Metal optimization(s) already. So are all the possible opts included aka don't needa invoke them in some way? 

Also have ideas for other integrations -- like using this fork of bitsandbytes https://github.com/rickardp/bitsandbytes/tree/portable (idk if the Apple optimizations are in that fairly current branch, https://github.com/rickardp/bitsandbytes/tree/apple-silicon-merge2), then there's Flash Metal Attention https://github.com/philipturner/metal-flash-attention

and a *highly* involved 1 -- are y'all aware of Tinygrad? https://github.com/tinygrad/tinygrad 
Though maybe it is easy if you know what you are doing, I'm an absolute pleb, not a coder by trade. Seems there's some automated functionalities but I couldn't tell for certain.
